Submit Homework 3 for Distributed Computing: Monitoring-enabled WordPress cluster with GTID-based MySQL replication using Ansible and Docker

DmitryFedoroff opened the pull request:
- Modularize MySQL replication logic into reusable task configure_replica.yml with GTID and health validation.
- Introduce cAdvisor container for Docker-level performance metrics.
- Provision Prometheus service with external scrape configuration for cAdvisor.
- Provision Grafana container with predefined Prometheus datasource.
- Automate WordPress container initialization with WP-CLI core installation.
- Refactor docker_setup to ensure idempotent installation of Docker, Compose and SDKs.
- Consolidate ansible.cfg and inventory.yml to support password-based connection with paramiko.
- Remove deprecated site.yml and static inventory/hosts.ini
- Improve playbooks modularity and role separation across playbook1.yml, playbook2.yml, and playbook3.yml

DmitryFedoroff committed:
build: Configure dependencies for container management

#comment Configure Ansible collection requirements for Docker-related modules ensuring community.docker availability

Affected: requirements.yml

DmitryFedoroff committed:
chore: Provide inventory and variable definitions

#comment Provide local connection settings, Python interpreter paths, and WordPress credentials for environment consistency

Affected: hosts.ini, local.yml, all.yml

DmitryFedoroff committed:
feat: Orchestrate site deployment

#comment Orchestrate Ansible roles for Docker, database, and WordPress to enable cohesive setup

Affected: site.yml

DmitryFedoroff committed:
feat: Establish Docker infrastructure

#comment Establish tasks for Docker Engine installation, Compose retrieval, and container orchestration with a dedicated network

Affected: main.yml, docker-compose.yml.j2

DmitryFedoroff committed:
feat: Configure database readiness

#comment Configure waiting mechanism for MySQL and specify credentials to ensure the service is ready before further steps

Affected: main.yml, db-config.yml.j2

DmitryFedoroff committed:
feat: Provision WordPress framework

#comment Provision tasks and templates for custom wp-config with essential database details and environment variables

Affected: main.yml, wp-config.php.j2

DmitryFedoroff committed:
feat: Configure Nginx routing

#comment Configure server block to route requests through PHP-FPM and serve the WordPress files from the correct document root

Affected: nginx.conf

DmitryFedoroff committed:
chore: Add screenshots folder for documentation

#comment Add graphical assets demonstrating deployment state including WordPress installation, admin console, public interface and running containers

Affected: screenshot_01.png, screenshot_02.png, screenshot_03.png, screenshot_04.png

DmitryFedoroff committed:
doc: Document deployment of containerized WordPress infrastructure

#comment Document project architecture, environment requirements, Ansible role structure, Docker service composition and startup procedure with embedded annotated screenshots for verification and reproducibility

Affected: README.md

DmitryFedoroff committed:
feat: Merge HW1 deployment pipeline for containerized WordPress infrastructure

This pull request finalizes Homework 1 for Distributed Computing by delivering a reproducible infrastructure-as-code solution for deploying WordPress using Ansible and Docker.

Highlights:
- Dockerized architecture for MySQL, WordPress (PHP-FPM), and Nginx.
- Ansible roles for Docker installation, service orchestration, and wp-config templating.
- Dynamic inventory and environment variable management.
- Custom Nginx configuration for PHP routing.
- Deployment validated via WordPress UI and CLI state (screenshots included).

DmitryFedoroff committed:
refactor: Update WordPress stack deployment for remote server compatibility and container stability

#comment Refactor WordPress stack deployment:

1. Remove legacy templates wp-config.php.j2 and db-config.yml.j2 to rely on built-in WordPress configuration via environment variables.
2. Configure Docker CE version 28.0.4 and docker-compose-plugin installation via apt with version validation, avoiding insecure curl-based downloads.
3. Relocate nginx.conf into role-specific files directory to follow Ansible structure conventions.
4. Eliminate redundant host_vars/local.yml and consolidate configuration under group_vars/all.yml.
5. Update MySQL, WordPress, and Nginx image tags to fixed stable versions for reproducibility and security.
6. Restrict MySQL service to internal Docker network by removing external port exposure.
7. Implement MySQL healthcheck and enforce startup dependency for WordPress container on database readiness.
8. Ensure persistent volumes are created for both database and application services.
9. Transition Ansible inventory to target remote Ubuntu host via SSH and update playbook to reflect new host group.
10. Maintain idempotency by avoiding shell or command tasks in favor of declarative Ansible modules.
11. Update deployment documentation and visual references via refreshed screenshots.

Affected: group_vars_all_yml, host_vars_local_yml, inventory_hosts_ini, requirements_yml, roles.db_setup.config_files_db_config_yml_j2, roles.db_setup.tasks_main_yml, roles.docker_setup.tasks_main_yml, roles.docker_setup.files_nginx_conf, roles.docker_setup.templates_docker_compose_yml_j2, roles.wordpress_setup.tasks_main_yml, roles.wordpress_setup.templates_wp_config_php_j2, screenshots_screenshot_01_png, screenshots_screenshot_02_png, screenshots_screenshot_03_png, screenshots_screenshot_04_png, site_yml

DmitryFedoroff committed:
refactor: Configure MySQL cluster with GTID replication and legacy data migration

#comment Configure multi-node MySQL cluster with one master and two replicas using GTID-based replication:

1. Replace single MySQL container with multi-node cluster including master and two slaves.
2. Create master.cnf, slave1.cnf and slave2.cnf with GTID, binlog, and replication directives.
3. Adjust docker-compose.yml to define db-master, db-slave1, db-slave2 with configuration volumes and healthchecks.
4. Integrate master/slave MySQL server configs into docker_setup role and copy them to remote host.
5. Detect legacy MySQL volume, run temporary container and dump all data into the new master.
6. Create replication user inside the master node using docker_container_exec.
7. Start replication on both slaves using CHANGE REPLICATION SOURCE TO with GTID-based auto-positioning.
8. Modify db_setup tasks to handle healthcheck, legacy container dump, and slave synchronization.
9. Refactor all.yml to include replication credentials and volumes for cluster migration.

Affected: group_vars/all.yml, docker_setup/tasks/main.yml, docker_setup/templates/docker-compose.yml.j2, docker_setup/files/master.cnf, docker_setup/files/slave1.cnf, docker_setup/files/slave2.cnf, db_setup/tasks/main.yml

DmitryFedoroff committed:
Merge pull request #2 from DmitryFedoroff/feat/hw2-distributed-computing-itmo

feat: Merge HW2 database clustering for MySQL GTID-based replication and legacy migration

DmitryFedoroff committed:
doc: Update README with GTID MySQL cluster architecture and migration flow

#comment Update documentation to reflect improvements in Homework 2 deployment pipeline:

1. Describe transition to distributed MySQL architecture with GTID replication (Master-Slave-Slave).
2. Document legacy volume detection and automated migration strategy via temporary container.
3. Outline Docker Compose network structure, healthchecks, and replication credentials setup.
4. Reflect infrastructure update to remote server with external Nginx port binding and container naming.
5. Clarify WordPress configuration without wp-config.php templating using environment variables.
6. Detail installation instructions including Python venv, Ansible Core, Galaxy collections, and SSH setup.
7. Include updated screenshots of admin UI, Docker containers, and WordPress landing page.
8. Present technical insights into failover, replication mechanics, and compose orchestration model.

Affected: README.md

DmitryFedoroff committed:
Merge pull request #3 from DmitryFedoroff/doc/update-readme-for-gtid-cluster-deployment

doc: Update README with GTID MySQL cluster architecture and migration flow

DmitryFedoroff committed:
refactor: Replace unsafe command executions with structured Docker exec calls

#comment Refactor MySQL replication setup by eliminating all shell-style command invocations:

1. Convert all mysql calls to argv format for safer container execution
2. Replace sh -c with explicit /bin/bash only where piping requires shell
3. Ensure all changes are restricted to docker_container_exec usage
4. Maintain idempotent logic and original orchestration flow

Affected: roles/db_setup/tasks/main.yml

DmitryFedoroff committed:
refactor: Replace unsafe command executions with structured Docker exec calls

#comment Refactor MySQL replication setup by eliminating all shell-style command invocations:

1. Convert all mysql calls to argv format for safer container execution
2. Replace sh -c with explicit /bin/bash only where piping requires shell
3. Ensure all changes are restricted to docker_container_exec usage
4. Maintain idempotent logic and original orchestration flow

Affected: roles/db_setup/tasks/main.yml

DmitryFedoroff committed:
refactor: Transition to declarative MySQL cluster management with community.mysql

#comment Refactor infrastructure code to replace imperative docker_container_exec usage with safe and idempotent community.mysql modules for MySQL user creation, data import, and replication setup. Include dependency on community.mysql, enable PyMySQL, switch to declarative dump/import, and rework docker-compose for external network compatibility.

Affected: requirements.yml, db_setup/tasks/main.yml, docker_setup/tasks/main.yml, docker_setup/templates/docker-compose.yml.j2

DmitryFedoroff committed:
refactor: Refine Docker and MySQL roles for idempotent secure replication

#comment Refactor installation and replication logic to increase idempotency, security and observability:

1. Collect package facts via apt and expose current docker‑ce and compose plugin versions for operators.
2. Verify version compliance against 5:28.0.4 and 2.20.0 thresholds and conditionally add Docker APT repo and GPG key.
3. Separate engine and plugin installation paths using docker_pkg_version and compose_plugin_min_version facts.
4. Skip redundant package upgrades when required versions already present to keep reruns fast and safe.
5. Recreate replication account with REQUIRE SSL and GTID‑safe grants to enforce encrypted channels.
6. Hard reset replica state on slaves and apply CHANGE PRIMARY with primary_ssl for auto‑positioned GTID replication.
7. Insert pause to allow catch‑up and assert IO and SQL thread health with explicit fail‑fast checks.
8. Validate wordpress_db existence on each replica and report Seconds_Behind_Source for troubleshooting.
9. Output informative debug messages to clarify play behaviour during dry‑runs and real executions.

Affected: docker_setup/tasks/main.yml, db_setup/tasks/main.yml

DmitryFedoroff committed:
Merge branch 'main' into feat/hw2-distributed-computing-itmo

DmitryFedoroff committed:
refactor: Integrate GTID replication and monitoring stack for distributed WordPress cluster

#comment Refactor WordPress deployment and monitoring infrastructure for Homework 3 of Distributed Computing course:

1. Modularize MySQL replication logic into reusable task configure_replica.yml with GTID and health validation.
2. Introduce cAdvisor container for Docker-level performance metrics.
3. Provision Prometheus service with external scrape configuration for cAdvisor.
4. Provision Grafana container with predefined Prometheus datasource.
5. Automate WordPress container initialization with WP-CLI core installation.
6. Refactor docker_setup to ensure idempotent installation of Docker, Compose and SDKs.
7. Consolidate ansible.cfg and inventory.yml to support password-based connection with paramiko.
8. Remove deprecated site.yml and static inventory/hosts.ini
9. Improve playbooks modularity and role separation across playbook1.yml, playbook2.yml, and playbook3.yml

Affected: ansible.cfg, group_vars/all.yml, inventory.yml, playbook1.yml, playbook2.yml, playbook3.yml, requirements.yml, roles/cadvisor/tasks/main.yml, roles/db_setup/tasks/configure_replica.yml, roles/db_setup/tasks/main.yml, roles/docker_setup/tasks/main.yml, roles/grafana/files/datasource.yml, roles/grafana/tasks/main.yml, roles/prometheus/files/prometheus.yml, roles/prometheus/tasks/main.yml, roles/wordpress_setup/tasks/main.yml, .ansible-lint.yaml, .github/workflows/ansible-lint.yml, site.yml, inventory/hosts.ini

SemenMartynov commented:
```
      ansible_connection: paramiko  # Обрабатывает пароль внутри Python-кода
```

В 4-й работе инвентори генерится автоматически, так что эту переменную стоит перенести в vars.

Падает.
```
TASK [Gathering Facts] **************************************************************************************************************************************************************************************
fatal: [carrier]: FAILED! => {"msg": "paramiko is not installed: No module named 'paramiko'"}
```

А, так, это модуль... Вообще, в зависимостях я его не вижу. И по коду (через поиск) тоже не вижу упоминания.

---

Упало тут
```
TASK [db_setup : Stop replica if running] *******************************************************************************************************************************************************************
fatal: [carrier]: FAILED! => {"changed": false, "msg": "unable to find /root/.my.cnf. Exception message: (2003, \"Can't connect to MySQL server on '172.18.0.5' ([Errno 111] Connection refused)\")"}
...ignoring

TASK [db_setup : Ensure replica configured for master] ******************************************************************************************************************************************************
fatal: [carrier]: FAILED! => {"changed": false, "msg": "unable to find /root/.my.cnf. Exception message: (2003, \"Can't connect to MySQL server on '172.18.0.5' ([Errno 111] Connection refused)\")"}
```

Я не лезу разбираться, т.к. не очень понял идею с paramiko. Ну и вижу там игнорирование ошибок -- это плохая практика, ошибки нужно решать (обрабатывать) а не игнорить...

DmitryFedoroff committed:
refactor: Consolidate Grafana provisioning and refine MySQL replication control logic

#comment Refactor monitoring and replication subsystems as follows:

1. Configure automated Grafana dashboard provisioning via dashboards.yml and a JSON-based metrics panel layout.
2. Refactor Grafana container mounting logic to include dashboards volume.
3. Improve readability of Grafana tasks by separating dashboard provisioning paths.
4. Handle edge-case failures during MySQL stopreplica stage using safe conditional logic.
5. Extract startreplica execution into dedicated handler for modular reuse.
6. Validate database existence on slaves after GTID replication is established.
7. Refactor docker_setup tasks to unify version check logic and simplify debug output.
8. Update group_vars to include structured Grafana dashboard path variable and correct WordPress admin email.
9. Adjust WP core install logic to support idempotent deployment and clearly interpret exit codes.

Affected: roles/grafana/files/datasource.yml, roles/grafana/files/dashboards.yml, roles/grafana/files/dashboard.json, roles/grafana/tasks/main.yml, roles/db_setup/tasks/configure_replica.yml, roles/db_setup/handlers/main.yml, roles/db_setup/tasks/main.yml, roles/docker_setup/tasks/main.yml, roles/wordpress_setup/tasks/main.yml, group_vars/all.yml, inventory.yml, requirements.yml

DmitryFedoroff commented:
> ```
>       ansible_connection: paramiko  # Обрабатывает пароль внутри Python-кода
Исправления сделаны. Удалил `ansible_connection: paramiko`. paramiko работал, но действительно было здесь лишним. Внесены правки `ansible-lint` и исправлен ряд сопутствующих, но важных замечаний.
![Мониторинг](https://github.com/user-attachments/assets/941c7c41-6c2a-40f6-b991-34115ce1f474)

SemenMartynov reviewed (approved):
Порядок, до мониторинга добрался.
